firedrake/pointcloud.py

```diff
@@ -0,0 +1,283 @@
+from firedrake.mesh import spatialindex
+from firedrake.dmplex import build_two_sided
+from firedrake.utils import cached_property
+from firedrake.petsc import PETSc
+from firedrake import logging
+from pyop2.mpi import MPI
+from pyop2.datatypes import IntType, ScalarType
+import numpy as np
+
+
+__all__ = ["PointCloud"]
+
+
+def syncPrint(*args, **kwargs):
+    """Perform a PETSc syncPrint operation with given arguments if the logging level is
+    set to at least debug.
+    """
+    if logging.logger.isEnabledFor(logging.DEBUG):
+        PETSc.Sys.syncPrint(*args, **kwargs)
+
+
+def syncFlush(*args, **kwargs):
+    """Perform a PETSc syncFlush operation with given arguments if the logging level is
+    set to at least debug.
+    """
+    if logging.logger.isEnabledFor(logging.DEBUG):
+        PETSc.Sys.syncFlush(*args, **kwargs)
+
+
+class PointCloud(object):
+    """Store points for repeated location in a mesh. Facilitates lookup and evaluation
+    at these point locations.
+    """
+    def __init__(self, mesh, points, *args, **kwargs):
+        """Initialise the PointCloud.
+
+        :arg mesh: A mesh object.
+        :arg points: An array of points.
+        """
+        self.mesh = mesh
+        self.points = np.asarray(points)
+        _, dim = points.shape
+        if dim != mesh.geometric_dimension():
+            raise ValueError("Points must be %d-dimensional, (got %d)" %
+                             (mesh.geometric_dimension(), dim))
+
+        # Build spatial index of processes for point location.
+        self.processes_index = self._build_processes_spatial_index()
+
+    @cached_property
+    def locations(self):
+        """Determine the process rank and element that holds each input point.
+
+        The location algorithm works as follows:
+        1. Query local spatial index for list of input points (`self.points`) to determine
+           if each is held locally.
+        2. For each point not found locally, query spatial index of processes to
+           identify those that may contain it. Location requests will be sent to each
+           process for the points they may contain.
+        3. Perform communication round so that each process knows how many points it will
+           receive from the other processes.
+        4. Perform sparse communication round to receive points.
+        5. Lookup these points in local spatial index and obtain result.
+        6. Perform sparse communication round to return results.
+        7. Process responses to obtain final results array. If multiple processes report
+           containing the same point, choose the process with the lower rank.
+
+        :returns: An array of (rank, cell number) pairs; (-1, -1) if not found.
+        """
+        rank_cell_pairs = self._locate_mesh_elements()
+        return np.array(rank_cell_pairs, dtype=IntType)
+
+    def evaluate(self, function):
```

Comment:
I'm not convinced that having an evaluate method here is wise. What happens if I want to use this in the Functional for an optimisation problem? What does the symbol look like for the derivative of this operator with respect to the point cloud? What does its adjoint look like?

I think a PointCloud is analogous to a FunctionSpace. We probably need PointFunction. Then you would have an `UFL.Delta(expr, cloud)` which would evaluate expr on the point cloud and gives us the right place to do all the derivatives and adjoints. 

I guess it would be OK to have an evaluate method as long as what it does is `return assemble(UFL.Delta(expr, self))`

Reply:
OK, let's punt on this for now (and remove the `evaluate` method). Instead a baby step in the right direction will be to make `Function.at` use the locations from here (rather than locating itself).